zebra-core/src/Zebra/Merge/Table.hs

```diff
@@ -96,156 +119,189 @@ peekHead input = do
+  Stream.map (uncurry Row) $
+    Stream.concat $
+    Stream.mapM (hoistEither . logicalPairs) $
+    hoist lift 
+      stream
+{-# INLINABLE streamStripedAsRows #-}
+
+-- merge streams in a binary tree fashion
+mergeStreamsBinary ::
      Monad m
-  => Maybe MaximumRowSize
-  -> Cons Boxed.Vector (Input m)
-  -> Map Logical.Value Int64
-  -> Stream (Of Logical.Table) (EitherT UnionTableError m) ()
-unionInput msize inputs0 sizes0 = do
-  (inputs1, sizes1) <- lift $ runStateT (traverse updateInput inputs0) sizes0
-  unless (Cons.all isClosed inputs1) $ do
-    let
-      drops =
-        takeExcessiveValues msize sizes1
-
-      inputs2 =
-        fmap (dropData drops) inputs1
-
-      maximums =
-        Cons.mapMaybe (maximumKey . inputData) inputs1
-
-    if Boxed.null maximums then
-      unionInput msize inputs2 sizes1
-    else do
-      Step values inputs3 <- lift $ unionStep (Boxed.minimum maximums) inputs2
+  => Cons Boxed.Vector (Stream (Of Row) (EitherT UnionTableError m) ())
+  -> Stream (Of Row) (EitherT UnionTableError m) ()
+mergeStreamsBinary kvss =
+  case Cons.length kvss of
+    0 ->
+      pure mempty
+
+    1 ->
+      fromJust $ Cons.index 0 kvss
+
+    2 -> do
+      let mergeS s1 s2 = remapStreamEnd $ Stream.merge s1 s2
+      Cons.foldl1 mergeS kvss
+
+    n -> do
       let
-        unyieldedSizes
-          = sizes1 `Map.difference` values
+        (kvss0, kvss1) = Boxed.splitAt (n `div` 2) $ Cons.toVector kvss
+        kvs0 = mergeStreamsBinary $ Cons.unsafeFromVector kvss0
+        kvs1 = mergeStreamsBinary $ Cons.unsafeFromVector kvss1
+      mergeStreamsBinary $ Cons.from2 kvs0 kvs1
+{-# INLINABLE mergeStreamsBinary #-}
 
-      Stream.yield $ Logical.Map values
-      unionInput msize inputs3 unyieldedSizes
-{-# INLINABLE unionInput #-}
 
-unionLogical ::
+remapStreamEnd :: Monad m => Stream (Of t) m (r,r) -> Stream (Of t) m r
+remapStreamEnd s = 
+  s >>= (\(a,b) -> return $ seq b a)
+{-# INLINABLE remapStreamEnd #-}
+
+  
+isPastMaxRowSize :: Maybe MaximumRowSize -> Int64 -> Bool
+isPastMaxRowSize = \case
+  Nothing ->
+    const False
+  Just size ->
+    (> unMaximumRowSize size)
+{-# INLINABLE isPastMaxRowSize #-}
+
+
+unionInputGroupBy ::
      Monad m
   => Schema.Table
   -> Maybe MaximumRowSize
-  -> Cons Boxed.Vector (Stream (Of Logical.Table) m ())
-  -> Stream (Of Logical.Table) (EitherT UnionTableError m) ()
-unionLogical schema msize inputs = do
-  Stream.whenEmpty (Logical.empty schema) $
-    unionInput msize (fmap (Input Map.empty . Just) inputs) Map.empty
-{-# INLINABLE unionLogical #-}
+  -> Cons Boxed.Vector (Stream (Of Striped.Table) m ())
+  -> Stream (Of Row) (EitherT UnionTableError m) ()
+unionInputGroupBy schema msize inputs0 = do
+  let 
+    compKey :: Row -> Row -> Bool
+    compKey a b = (rowKey a) == (rowKey b)
+
+  Stream.mapMaybeM (hoistEither . mergeRows msize schema) $
+    Stream.filter (/= []) $
+    Stream.mapped Stream.toList $
+    Stream.groupBy compKey $
```

Comment:
I think what's happening here is that you're merging the streams using a binary tree, such that the final stream is ordered on the row key, then using `groupBy` to get all the rows which contain the key as a stream or list of vector, then re-turning it into a binary tree using the function `mergeValues`.

So I think this is a bit round about. It should be possible to merge the streams themselves in a binary fashion (and it would make doing it in parallel easier I believe).